packages/flutter/lib/src/material/animated_raster.dart

```diff
@@ -0,0 +1,202 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// TODO(jonahwilliams): this is temporarily a private library until we determine
+// a reasonable public API.
+part of page_transitions_theme;
+
+// Foundational widgets for the rasterized zoom page transition.
+
+/// A delegate that controls what is painted during an animation in the [_AnimatedRaster]
+/// widget.
+@immutable
+abstract class _AnimatedRasterDelegate {
+  /// const constructor so that subclasses can be const.
+  const _AnimatedRasterDelegate();
+
+  /// Whether [paint] will paint anything on the current frame.
+  ///
+  /// This is used to optimize painting and delay rasterization if a child
+  /// would not be painted. One example would be an animated opacity, if the
+  /// first frame is fully transparent then rasterization can be delayed until
+  /// the second frame.
+  @protected
+  bool useRaster(Animation<double> animation) => true;
+
+  /// Paint the children of this render object.
+  ///
+  /// The provided [image] contains a GPU resident texture of the children of
+  /// this render object.
+  ///
+  /// The [pixelRatio] is the ratio of pixels in [image] to logical pixels.
+  @protected
+  void paintRaster(PaintingContext context, ui.Image image, double pixelRatio, Animation<double> animation);
+
+  /// Paint the children of this render object.
```

Comment:
It strikes me as a little odd that this paint is only used when `useRaster` is false, but isn't used to paint the children for the initial rasterization if `useRaster` is true...

Reply:
I could see us breaking this into multiple simpler ROs which might alieviate the problem - which is that this RO + delegate need to do too much - that being said, one issue is that for this API to be flexible enough we should have access to the resulting image somehow. I'm not sure how to do that with composition.

Reply:
maybe _all_ the delegate does is handle drawing the image, the notifier is separate, and then the parent which controls the notifier chooses whether to draw with any effects? That might let us combine the two animations. I'll investigate that a bit

Reply:
Maybe instead of having a delegate, we have an abstract class as public that does the `ideal` case above, and a private concrete implementation that does the `reality` case?